pkg/controller/datasource-controller.go

```diff
@@ -121,24 +126,88 @@ func NewDataSourceController(mgr manager.Manager, log logr.Logger, installerLabe
 		scheme:          mgr.GetScheme(),
 		log:             log.WithName(dataImportControllerName),
 		installerLabels: installerLabels,
+		dsMap:           make(map[types.NamespacedName][]reconcile.Request),
 	}
 	DataSourceController, err := controller.New(dataImportControllerName, mgr, controller.Options{Reconciler: reconciler})
 	if err != nil {
 		return nil, err
 	}
-	if err := addDataSourceControllerWatches(mgr, DataSourceController, log); err != nil {
+	if err := addDataSourceControllerWatches(mgr, DataSourceController, &reconciler.dsMap); err != nil {
 		return nil, err
 	}
 	log.Info("Initialized DataSource controller")
 	return DataSourceController, nil
 }
 
-func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller, dsMap *dataSourceMap) error {
 	if err := cdiv1.AddToScheme(mgr.GetScheme()); err != nil {
 		return err
 	}
-	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return updateDataSourceMap(e.Object, dsMap, false) },
+			DeleteFunc: func(e event.DeleteEvent) bool { return updateDataSourceMap(e.Object, dsMap, true) },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return !sameDataSourcePvc(e.ObjectOld, e.ObjectNew) &&
+					updateDataSourceMap(e.ObjectOld, dsMap, true) &&
+					updateDataSourceMap(e.ObjectNew, dsMap, false)
+			},
+		},
+	); err != nil {
+		return err
+	}
+
+	mapToDataSource := func(obj client.Object) []reconcile.Request {
+		return (*dsMap)[types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}]
+	}
+	if err := c.Watch(&source.Kind{Type: &cdiv1.DataVolume{}},
+		handler.EnqueueRequestsFromMapFunc(mapToDataSource),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			// Only DV status phase update is interesting to reconcile
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				dvOld, okOld := e.ObjectOld.(*cdiv1.DataVolume)
+				dvNew, okNew := e.ObjectNew.(*cdiv1.DataVolume)
+				return okOld && okNew && dvOld.Status.Phase != dvNew.Status.Phase
+			},
+		},
+	); err != nil {
```

Comment:
Took me a while to figure out what is going on here.  Looks like the main reason for the map is to associate a PVC to a list of DataSources.  Fine.  Turns out that kubernetes informers have a standard way to build/maintain such indexes.  And controller runtime has their own way to manage them.  Best example I could find is here:  https://github.com/kubernetes-sigs/controller-runtime/blob/b704f447ea7c8f7059c6665143a4aa1f6da28328/pkg/client/example_test.go#L249-L267

You should be able to call `mgr.GetFieldIndexer()` in `NewDataSourceController` function to set everything up.  Then `client.List` can be used.

Alternatively, just list all DataSources in the DataVolume namespace to figure out which one to queue.  They should be cached.  At least that is easy to understand even if performance isn't optimal.

Reply:
The second alternative you just suggested is the first commit in this PR, which is quite expansive. Give it a look.
I'll try to switch to informers.

Reply:
Fixed